File: agc023/b/main.rb
require 'pp'
N = gets.chomp.to_i
m = []
good_m = []
N.times do
  m << gets.chomp.chars
  good_m << Array.new(N)
end


N.times do |i|
  N.times do |j|
    v = 0
    (1..N-1).each do |k|
      break if m[(i+k)%N][j] != m[i][(j+k)%N]
      v += 1
    end
    good_m[i][j] = v
  end
end

# pp good_m

ans = 0
N.times do |i|
  N.times do |j|
    good = true
    N.times do |k|
      unless good_m[(i+k)%N][(j+k)%N] >= N - 1 - k
        good = false
        break
      end
    end
    ans += 1 if good
  end
end


puts ans

File: agc023/b/main.go
package main

import (
	"fmt"
)

func main() {
	var (
		n int
	)
	fmt.Scan(&n)

	m := make([]string, n)
	goodM := make([][]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&m[i])
		goodM[i] = make([]int, n)
	}

	// fmt.Println(m)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			v := 0
			for k := 1; k < n; k++ {
				if m[(i+k)%n][j] != m[i][(j+k)%n] {
					break
				}
				v++
			}
			goodM[i][j] = v
		}
	}

	// fmt.Println(goodM)

	ans := 0

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			good := true
			for k := 0; k < n; k++ {
				if goodM[(i+k)%n][(j+k)%n] < n-1-k {
					good = false
					break
				}
			}
			if good {
				ans++
			}
		}
	}

	fmt.Printf("%d\n", ans)
}
